File: 310971/Khulup/task_310_rest/app/services/user_service.py
from app.repositories.user_repository import UserRepository
from app.schemas.user import UserCreate
from app.models.user import User
from typing import Optional

class UserService:
    def __init__(self, repo: UserRepository):
        self.repo = repo

    def create_user(self, user_data: UserCreate) -> User:
        return self.repo.add(user_data)

    def get_user(self, user_id: int) -> Optional[User]:
        return self.repo.get_by_id(user_id)

    def list_users(self) -> list[User]:
        return self.repo.list_users()

    def delete_user(self, user_id: int) -> bool:
        return self.repo.delete(user_id)

    def update_user(self, user_id: int, user_data: UserCreate) -> Optional[User]:
        return self.repo.update(user_id, user_data)

File: 351001/polozkov/src/main/java/org/polozkov/service/user/UserService.java
package org.polozkov.service.user;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.polozkov.dto.user.UserRequestTo;
import org.polozkov.dto.user.UserResponseTo;
import org.polozkov.entity.issue.Issue;
import org.polozkov.entity.user.User;
import org.polozkov.exception.BadRequestException;
import org.polozkov.mapper.user.UserMapper;
import org.polozkov.repository.user.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import java.util.ArrayList;
import java.util.List;

@Service
@Validated
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final UserMapper userMapper;

    public List<UserResponseTo> getAllUsers() {
        return userRepository.findAll().stream()
                .map(userMapper::userToResponseDto)
                .toList();
    }

    public UserResponseTo getUser(Long id) {
        User user = userRepository.getById(id);
        return userMapper.userToResponseDto(user);
    }

    public UserResponseTo createUser(@Valid UserRequestTo userRequest) {
        if (userRepository.findByLogin(userRequest.getLogin()).isPresent()) {
            throw new BadRequestException("User with login " + userRequest.getLogin() + " already exists");
        }

        User user = userMapper.requestDtoToUser(userRequest);

        user.setIssues(new ArrayList<>());

        User savedUser = userRepository.save(user);
        return userMapper.userToResponseDto(savedUser);
    }

    public UserResponseTo updateUser(@Valid UserRequestTo userRequest) {
        User existingUser = userRepository.getById(userRequest.getId());

        User user = userMapper.updateUser(existingUser, userRequest);

        user.setIssues(existingUser.getIssues());

        User updatedUser = userRepository.update(user);
        return userMapper.userToResponseDto(updatedUser);
    }

    public void deleteUser(Long id) {
        User user = userRepository.getById(id);

        if (!user.getIssues().isEmpty()) {
            throw new BadRequestException("Cannot delete user with existing issues. Delete issues first.");
        }

        userRepository.deleteById(id);
    }

    public User getUserById(Long id) {
        return userRepository.getById(id);
    }

    public void addIssueToUser(Long userId, Issue issue) {
        User user = userRepository.getById(userId);
        user.getIssues().add(issue);
        userRepository.update(user);
    }
}